xgcm/grid.py

```diff
@@ -1268,19 +1268,49 @@ def set_metrics(self, key, value):
                 f"Metric axes {axes_not_found!r} not compatible with grid axes {tuple(self.axes)!r}"
             )
 
-        # initialize empty list
-        self._metrics[metric_axes] = []
-        for metric_varname in _maybe_promote_str_to_list(value):
-            if metric_varname not in self._ds:
-                raise KeyError(
-                    f"Metric variable {metric_varname} not found in dataset."
-                )
-            # resetting coords avoids potential broadcasting / alignment issues
-            metric_var = self._ds[metric_varname].reset_coords(drop=True)
+        metric_value = _maybe_promote_str_to_list(value)
+        existing_metric_axes = set(k for k in self._metrics.keys())
 
-            # TODO: check for consistency of metric_var dims with axis dims
-            # check for duplicate dimensions among each axis metric
-            self._metrics[metric_axes].append(metric_var)
+        if metric_axes in existing_metric_axes:
+            value_exist = self._metrics.get(metric_axes)
+            for metric_varname in metric_value:
+                if metric_varname not in self._ds:
+                    raise KeyError(
+                        f"Metric variable {metric_varname} not found in dataset."
+                    )
+                else:
+                    value_new = self._ds[metric_varname].reset_coords(drop=True)
+                    # check for duplicate values
+                    if any(value_new.equals(ve) for ve in value_exist):
+                        # go through each existing value until data array with matching dimensions is selected
+                        for idx, ve in enumerate(value_exist):
+                            # double check if dimensions match
+                            if set(value_new.dims) == set(ve.dims):
+                                warnings.warn(
```

Comment:
This should only be triggered if overwrite is true? I think you just need to move it a line down under `if overwrite:`. Could you add this to the tests,  and check if this is triggered only if the value is actually overwritten?

Reply:
So right now the flow is: 
1) check if there is input which matches an existing metric (Line 1284)
2) cycle through existing metrics to find the match 
3) check if the dimensions between input and existing metric match 
4) replace if the value of overwrite = True
Just to clarify would you want to switch steps 3 and 4?

Reply:
Oh no. I am just concerned about the warning. 

Currently you have somthing like this:
```python
if set(value_new.dims) == set(ve.dims):
    warn(... replacing...).# this will be triggered regardless of whether overwrite is true or false
    if overwrite:
        # replace metric
    else:
       raise error
```

What I suggest is simply to move the warning into the if statement. Otherwise you will get a very confusing output for overwrite false. It will basically say: "I am overwriting" (which it isnt) and then fail.

Reply:
What I would do here is again start with the test: Make a test that checks that this warning is raised when overwrite=True, and more importanty: Make sure it is **not** raised when overwrite=False. I believe this will fail with the current code, and then pass when you move the warning.